tests/integration/s2n_test_constants.py

```diff
@@ -137,3 +137,56 @@
 
 DEFAULT_CLIENT_CERT_PATH = TEST_CERT_DIRECTORY + "rsa_2048_sha256_client_cert.pem"
 DEFAULT_CLIENT_KEY_PATH = TEST_CERT_DIRECTORY + "rsa_2048_sha256_client_key.pem"
+
+TEST_SNI_CERT_DIRECTORY="../pems/sni/"
+# Server certificates used to test matching domain names client with server_name
+# ( cert_path, private_key_path, domains[] )
+SNI_CERTS = {
+    "alligator" : ( TEST_SNI_CERT_DIRECTORY + "alligator_cert.pem", TEST_SNI_CERT_DIRECTORY + "alligator_key.pem",
+        ["www.alligator.com"]),
+    "alligator_ecdsa" : ( TEST_SNI_CERT_DIRECTORY + "alligator_ecdsa_cert.pem", TEST_SNI_CERT_DIRECTORY +
+        "alligator_ecdsa_key.pem", ["www.alligator.com"]),
+    "beaver"    : ( TEST_SNI_CERT_DIRECTORY + "beaver_cert.pem", TEST_SNI_CERT_DIRECTORY + "beaver_key.pem",
+        ["www.beaver.com"]),
+    "many_animals" : (TEST_SNI_CERT_DIRECTORY + "many_animal_sans_rsa_cert.pem", TEST_SNI_CERT_DIRECTORY + "many_animal_sans_rsa_key.pem",
+        ["www.catfish.com",
+         "www.dolphin.com",
+         "www.elephant.com",
+         "www.falcon.com",
+         "www.gorilla.com",
+         "www.horse.com",
+         "www.impala.com",
+         # "Simple hostname"
+         "Jackal",
+         "k.e.e.l.b.i.l.l.e.d.t.o.u.c.a.n",
+         # SAN on this cert is actually "ladybug.ladybug"
+         # Verify case insensitivity works as expected.
+         "LADYBUG.LADYBUG",
+         "com.penguin.macaroni"
+        ])
+}
+
+# Test cases with certificates to load into s2nd and expected behavior
+# ( certificates_to_use[], (domain_name, expected_certificate, expected_domain_match, client_cipher)[]  )
+SNI_CERT_TEST_CASES = [
+    ([SNI_CERTS["alligator"], SNI_CERTS["beaver"], SNI_CERTS["alligator_ecdsa"]],
+        [("www.alligator.com", SNI_CERTS["alligator"], True, "ECDHE-RSA-AES128-SHA"),
+        ("www.beaver.com", SNI_CERTS["beaver"], True,  "ECDHE-RSA-AES128-SHA"),
+        # This is a mismatch, expect the first cert added is selected.
+        ("not.a.match", SNI_CERTS["alligator"], False, "ECDHE-RSA-AES128-SHA"),
+        # No SNI sent at all. expect the first cert added is selected.
+        (None, SNI_CERTS["alligator"], False, "ECDHE-RSA-AES128-SHA"),
+        # We have two certificates that match www.alligator.com but the client
+        # only supports ECDSA. The ECDSA cert should be served
+        ("www.alligator.com", SNI_CERTS["alligator_ecdsa"], True, "ECDHE-ECDSA-AES128-SHA"),
+        # Client supports mixed auth types, expect we negotiate the higher priority cipher
+        # and select the correct cert
+        ("www.alligator.com", SNI_CERTS["alligator_ecdsa"], True, "ECDHE-RSA-AES128-SHA:ECDHE-ECDSA-AES128-SHA")]),
```

Comment:
should the test take into account server preferences as well?

Reply:
We could, but I think we can get the coverage we need by assuming we use `test_all` server preferences and then alter the client side ciphers as needed. For example if we want "server should have a preferred ecdsa match" scenario we can make the client ciphers: `ECDHE-RSA-AES128-SHA:ECDHE-ECDSA-AES128-SHA` . If we want "server should have a preferred rsa match" we can make the client ciphers `ECDHE-RSA-AES128-SHA:ECDHE-ECDSA-AES128-SHA256`